Adding ability to castle, not yet passing all tests

hathcode opened the pull request.

cbohara commented on app/models/king.rb:
```diff
@@ -1,8 +1,12 @@
+    (normal_valid_move?(x,y) && super(x,y)) || can_castle?(x,y)
+  end
+
+  # ensures king piece can only move 1 square in any direction
```
`def valid_move?(x,y)
    # checks critera for valid_move? that apply to all pieces and returns false if fails
    return false unless super(x,y)
    x_diff = x_diff(x)
    y_diff = y_diff(y)
    # returns true if moving only 1 square any direction
    if (x_diff <= 1) && (y_diff <= 1)
      return true
    end
    # returns false if not moving only 1 square any direction
    return false
  end`

hathcode replied:
this is what was there before my changes:
class King < Piece
  # ensures king piece can only move 1 square in any direction
  def valid_move?(x,y)
    # checks critera for valid_move? that apply to all pieces and returns false if fails
    return false unless super(x,y)
    x_diff = x_diff(x)
    y_diff = y_diff(y)
    # returns true if moving only 1 square any direction
    if (x_diff <= 1) && (y_diff <= 1)
      return true
    end
    # returns false if not moving only 1 square any direction
    return false
  end

I moved the logic you mentioned to "normal valid move" and adjusted the tests for that as well.  It's still there, just rearranged to allow for castling.

cbohara replied:
Ooops! I clicked on the little accordion button on the side left navigation and got to see that that bit of code wasn't taken out.  Sorry about that!

jacobwgillespie commented on app/models/king.rb:
```diff
@@ -1,8 +1,12 @@
-    # checks critera for valid_move? that apply to all pieces and returns false if fails
-    return false unless super(x,y)
+    # checks critera for valid_move? that apply to all pieces and returns false if fails unless can_castle
+    (normal_valid_move?(x,y) && super(x,y)) || can_castle?(x,y)
```
The order should switch so that `super(x, y)` is called before `normal_valid_move?` as that preserves what happened before:

```ruby
(super(x, y) && normal_valid_move?(x,y)) || can_castle?(x,y)
```

You could, BTW, keep the old `valid_move?` and just add the castle check as another condition:

```ruby
  def valid_move?(x, y)
    # returns true if moving only 1 square any direction
    return true if (x_diff(x) <= 1) && (y_diff(y) <= 1)

    # returns true if the king can castle
    return true if can_castle?(x, y)

    return false
  end
```

hathcode replied:
Ahh, I was wondering about that Jacob,  I'll fix super and try that, and maybe try the other way as well.

jacobwgillespie commented on app/models/king.rb:
```diff
@@ -13,6 +17,35 @@ def valid_move?(x,y)
+
+
+  def move_to(x,y)
+    castle! if can_castle?(x,y)
```
You might want to rearrange this so that the parent `move_to` is called before calling `castle!` to ensure it succeeds.  Since `valid_move?` calls `can_castle?`, it should be safe to call `super` first.  Something like:

```ruby
def move_to(x, y)
  success = super(x, y)
  if success
    castle! if can_castle?(x, y)
  end
end
```

Also, if you call just plain `super` it doesn't pass the `x` and `y` parameters (IIRC), so you may need to use `super(x, y)` here.

hathcode replied:
Since we have to ensure that the path for the rook is not blocked (nothing between the rook and the king), won't this cause an error if the king is moved before the rook?

jacobwgillespie commented on app/models/king.rb:
```diff
@@ -13,6 +17,35 @@ def valid_move?(x,y)
 
+
+  def can_castle?(x,y)
+    updated_at == created_at && \
```
This method is a little complex - you might consider breaking it into smaller named methods like say `never_moved?` and `castle_requested?`, something like (untested pseudocode):

```ruby
class Piece < ActiveRecord::Base
  def never_moved?
    created_at == updated_at
  end
end

class King < Piece
  def castle_requested?(x, y)
    never_moved? && (x == 6 || x == 2)
  end
end
```

Then `can_castle?` can be a combination of smaller named methods and be easier to understand / test.

Also, as noted below on `castle_rook`, you'll want to take into account the case where the first rook you find isn't the castling rook, like if the castle direction was like so:

```
        ---->
R _ _ _ K _ _ R
```

You'll want to make sure to check / move the correct rook based on the castle direction, as a `.first` might pick up the rook on the left when the one on the right is supposed to move.

jacobwgillespie commented on app/models/king.rb:
```diff
@@ -13,6 +17,35 @@ def valid_move?(x,y)
+
+
+  def castle_rook
+    @castle_rook = game.pieces.where(x_position: 0, type: Rook).first || game.pieces.where(x_position: 7, type: Rook).first
```
This returns the first rook found, but not necessarily the correct rook as that is dependent on the castling direction.  This also returns *any* rook rather than only rooks of a friendly color.  Also, I believe the value of the `type` column will be a string - Rails may be smart enough to translate the class into a string, but it might be best to make it explicit:

```
def castle_rooks
  game.pieces.where(x_position: x_position, player_id: player_id, type: 'Rook')
end
```

That would return all rooks on the king's row that share the same player (color).  Combined with other checks inside `can_castle?`, you could pick which of those rooks (if any) are valid for castling.

hathcode replied:
I had it as 'Rook' at first, but it threw an error "unknown method or variable".  I'll try it again and double check.  
On def castle_rook, my thinking was that even though it would return any rook that might be in that x position, the other checks would narrow that down to the correct rook via "updated at = created at" confirming that the rook hadn't moved, and the y-position criteria verifying that it was in the same row as the king being moved.  Guess I had the right idea, just wasn't quite there!  I'll refactor with your suggestions and continue testing.  Thanks!

jacobwgillespie commented on spec/models/king_spec.rb:
```diff
@@ -2,82 +2,175 @@
       game = Game.create(white_user_id: 0, black_user_id: 1)
       king = King.create(x_position: 4, y_position: 0, game_id: game.id)
-      expect(king.valid_move?(5,0)).to eq true
+      expect(king.normal_valid_move?(5,0)).to eq true
```
The tests should still reference `valid_move?` not `normal_valid_move?` as that's what the rest of the app will be using.  In this case, `normal_valid_move?` is an "internal" method, part of how king's implement `valid_move?`.

hathcode commented:
I cleaned up the code and implemented the suggested changes (except didn't try 'rook' instead of Rook in the castle_rook method).  Still getting the same two tests not passing as before :-/  I'll keep plugging along.

hathcode commented:
Update:
So when i test the can_castle method, expect true but get false.
When I test the castle_queenside (or kingside) method, it returns nil, so it seems the problem is here, and causing the can_castle method to return false when it is called. Still working on it.

hathcode commented:
After tons of testing, it seems the problem is with the castle_rook method.  It is returning nil instead of the correct rook instance.  I've got some other things I need to do this afternoon, but if anyone has any  ideas how to get castle_rook fixed, I'm all eyes!!

hathcode commented:
Fixed the problem.  I still need to test "move_to!" in the king.rb file, and make sure the move is actually executed correctly, but the rest of the methods pass!

hathcode commented:
Seem to be having trouble with this bit of code:

  def castle!
    castle_rook.move_to!((castle_rook.x_position - (castle_rook.x_position - x_position).abs + 1), y_position)
  end

The formula that calculates the x_position checks (I verified it on a separate test), but the rook is not moving in the tests.  I can get the king to move if I use this for "move_to!"...

def move_to(x, y)
  success = super(x, y)
  if success
    castle! if can_castle?(x, y)
  end
end

..but still can't get the rook to move.  It seems a hang-up may be that when the rook is checking for path_blocked? that it sees the king already moved and returns true. I couldn't even get the rook to move by hard-coding the coordinates in the castle! method :-/

Suggestions are welcome!

cbohara commented on app/models/king.rb:
```diff
@@ -1,18 +1,53 @@
+    castle_rook.move_to!((castle_rook.x_position - (castle_rook.x_position - x_position).abs + 1), y_position)
+  end
+
+  def move_to!(x,y)
```
Hey Kevin! I'm a little confused by the move_to! method being in the king model.  I'm curious what this method is supposed to return.  I think on line 46 it should say `return castle! if can_castle?(x,y)`.  Right now this method only returns `super(x,y)` which I'm not sure is your intention.

hathcode replied:
The idea is that the moveto in the king model allows you to call the castle method within it.  The castle method should move the rook, move-to moves the king.  Here is what I have in my copy of the king.rb move_to!

  def move_to!(x,y)
    castle! if can_castle?(x,y)
    super(x,y)
  end

I played around with move_to a bunch before I pushed, so I may not have put all the code back in it before I pushed the last time.

jacobwgillespie replied:
It can't return after `castle!` because `castle!` only moves the rook - this still relies on `super` to move the king.

jacobwgillespie commented on app/models/king.rb:
```diff
@@ -1,18 +1,53 @@
+
+  #determines if there is a rook to castle with
+  def castle_rook
+    @castle_rook = game.pieces.where(y_position: y_position, user_id: user_id, type: 'Rook').first 
```
This method returns the first rook, not necessarily the correct rook.  This is causing some of your test failure.  Specifically, the first rook (`.first`), if neither have been moved, is always the queenside one, however your test for kingside castle fails because this method doesn't ever return the kingside rook.  You will want to either update this method to support choosing which rook you want, split it into two methods for each side, or refactor the logic into the other methods, based on what makes sense (to you, it's kinda subjective). 

👍

hathcode replied:
I figured it would at least work for kingside if there was no queenside rook so I could verify that it works in that limited case.  I'll work on it some more.  Thanks Jacob!

jacobwgillespie commented:
> It seems a hang-up may be that when the rook is checking for path_blocked? that it sees the king already moved and returns true.

You may want to change the logic so that either the king or the rook are manually moved (without `move_to!`) after you have validated that the castling is okay, that way you can bypass the `path_blocked?` check.  Essentially that means using `update_attributes` directly on either the rook or the king (or both, if you're sure that the castling is okay) after you've validated `can_castle?`

hathcode commented:
Ok, I suspected it was something like that.  I'll keep tweaking it and see if I can get it to work.  Thanks!
